lib/sort-order-selector/index.tsx

```diff
@@ -0,0 +1,122 @@
+import { recordEvent } from '../state/analytics/middleware';
+
+import * as S from '../state';
+import * as T from '../types';
+
+type StateProps = {
+  shouldDisplay: boolean;
+  sortReversed: boolean;
+  sortType: T.SortType;
+};
+
+type DispatchProps = {
+  setSortType: (sortType: T.SortType) => any;
+  toggleSortOrder: () => any;
+};
+
+type Props = StateProps & DispatchProps;
+
+export const SortOrderSelector: FunctionComponent<Props> = ({
+  shouldDisplay,
+  setSortType,
+  sortReversed,
+  sortType,
+  toggleSortOrder,
+}) => {
+  const changeSort = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    const selectedOption = event.currentTarget.selectedOptions[0];
+    const selectedSortType = selectedOption.dataset.order;
+    const selectedSortReversed = selectedOption.dataset.reversed;
+
+    if (selectedSortType !== sortType) {
+      setSortType(selectedSortType as T.SortType);
+    }
+    if ((selectedSortReversed === 'false') === sortReversed) {
+      toggleSortOrder();
+    }
+    recordEvent('list_sortbar_mode_changed', {
+      description: selectedOption.text,
+    });
+  };
+
+  const sortTypes = [
+    {
+      label: 'Name: A-Z',
+      id: 'alphabetical',
+      order: 'alphabetical',
+      reversed: false,
+    },
+    {
+      label: 'Name: Z-A',
+      id: 'alphabetical-reversed',
+      order: 'alphabetical',
+      reversed: true,
+    },
+    {
+      label: 'Created: Newest',
+      id: 'creationDate',
+      order: 'creationDate',
+      reversed: false,
+    },
+    {
+      label: 'Created: Oldest',
+      id: 'creationDate-reversed',
+      order: 'creationDate',
+      reversed: true,
+    },
+    {
+      label: 'Modified: Newest',
+      id: 'modificationDate',
+      order: 'modificationDate',
+      reversed: false,
+    },
+    {
+      label: 'Modified: Oldest',
+      id: 'modificationDate-reversed',
+      order: 'modificationDate',
+      reversed: true,
+    },
+  ];
+
+  return (
+    <Fragment>
+      {shouldDisplay && (
+        <div className="sort-order-selector theme-color-fg-dim theme-color-border">
+          <label htmlFor="sort-selection">Sort by</label>
+          <select
+            id="sort-selection"
+            value={sortType + ' ' + sortReversed.toString()}
+            onChange={changeSort}
+          >
+            {sortTypes.map((type) => (
+              <option
+                key={type.id}
+                value={type.id}
+                data-order={type.order}
+                data-reversed={type.reversed}
+              >
+                {type.label}
+              </option>
+            ))}
+          </select>
+        </div>
+      )}
+    </Fragment>
+  );
+};
+
+const mapStateToProps: S.MapState<StateProps> = (state) => ({
+  shouldDisplay: state.ui.filteredNotes.length > 0,
+  sortReversed: state.settings.sortReversed,
+  sortType: state.settings.sortType,
+});
+
+const mapDispatchToProps: S.MapDispatch<DispatchProps> = {
+  setSortType,
+  toggleSortOrder,
```

Comment:
calling two separate Redux actions introduces another partial update into our app. that is, if we change both sort properties then we're asking React to update after the first Redux dispatch and then again after the second. that intervening time represents a state that was never requested by the customer.

should we examine the `setSortType` action and give it the ability to also change the reversing of the search? if we did that, add an optional `reversed: boolean` property to the action, then we could perform the update in one go and we could eliminate the conditionals at the same time.

Reply:
Very cool! I do end up checking for `undefined` a bit, not sure if that's the right way to do this. https://github.com/Automattic/simplenote-electron/pull/2542/commits/3e8bac4c1d38ab1bf7ca5f7ee9d8bd09fde9222b